From 1e39c587a271ca1180dd28e9c92667c8dfa7619b Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:37:47 -0400
Subject: [ethernet] i40e: Don't disable SR-IOV when VFs are assigned

Message-id: <1409658042-22689-107-git-send-email-sassmann@redhat.com>
Patchwork-id: 90301
O-Subject: [RHEL7.1 PATCH 106/281] i40e: Don't disable SR-IOV when VFs are assigned
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Mitch Williams <mitch.a.williams@intel.com>

When VFs are assigned to active VMs and we disable SR-IOV out from under them,
bad things happen. Currently, the VM does not crash, but the VFs lose all
resources and have no way to get them back.

Add an additional check for when the user is disabling through sysfs, and add a
comment to clarify why we check twice.

Change-ID: Icad78eef516e4e1e4a87874d59132bc3baa058d4
Signed-off-by: Mitch Williams <mitch.a.williams@intel.com>
Tested-by: Sibai Li <sibai.li@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 9e5634dfb40c06c87f4a30a544f618f1c12897d4)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c b/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
index 82e7abf..4d21956 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
@@ -842,6 +842,10 @@ void i40e_free_vfs(struct i40e_pf *pf)
  kfree(pf->vf);
  pf->vf = NULL;
 
+ /* This check is for when the driver is unloaded while VFs are
+  * assigned. Setting the number of VFs to 0 through sysfs is caught
+  * before this function ever gets called.
+  */
  if (!i40e_vfs_are_assigned(pf)) {
   pci_disable_sriov(pf->pdev);
   /* Acknowledge VFLR for all VFS. Without this, VFs will fail to
@@ -978,7 +982,12 @@ int i40e_pci_sriov_configure(struct pci_dev *pdev, int num_vfs)
  if (num_vfs)
   return i40e_pci_sriov_enable(pdev, num_vfs);
 
- i40e_free_vfs(pf);
+ if (!i40e_vfs_are_assigned(pf)) {
+  i40e_free_vfs(pf);
+ } else {
+  dev_warn(&pdev->dev, "Unable to free VFs because some are assigned to VMs.\n");
+  return -EINVAL;
+ }
  return 0;
 }
 
-- 
1.7.1